src/content/profile-data.js

```diff
@@ -527,3 +527,285 @@ export function getTracingMarkers(thread, markers) {
+
+      // Insert them all into the empty row.
+      emptyTimingRow.stack.splice(emptyTimingIndex, 0, ...inRangeTimings.stack);
+      emptyTimingRow.start.splice(emptyTimingIndex, 0, ...inRangeTimings.start);
+      emptyTimingRow.end.splice(emptyTimingIndex, 0, ...inRangeTimings.end);
+
+      emptyTimingIndex = childTimingIndex;
+    }
+  }
+
+  // Set any platform stacks to -1
+  for (let depth = 0; depth < stackTimingByDepth.length; depth++) {
+    const timingRow = stackTimingByDepth[depth];
+    for (let i = 0; i < timingRow.stack.length; i++) {
+      const stackIndex = timingRow.stack[i];
+      const funcStackIndex = stackIndexToFuncStackIndex[stackIndex];
+      const funcIndex = funcStackTable.func[funcStackIndex];
+      if (!thread.funcTable.isJS[funcIndex]) {
+        timingRow.stack[i] = -1;
+      }
+    }
+  }
+
+  // Pre-emptively merge together consecutive platform stacks in the same row to minimize
+  // operations in the collapsing function.
+  for (let depth = 0; depth < stackTimingByDepth.length; depth++) {
+    const timingRow = stackTimingByDepth[depth];
+    for (let bIndex = 1; bIndex < timingRow.length; bIndex++) {
+      const aIndex = bIndex - 1;
+      const stackA = timingRow[aIndex];
+      const stackB = timingRow[bIndex];
+      if (stackA === -1 && stackB === -1) {
+        timingRow.start.splice(bIndex, 1);
+        timingRow.stack.splice(bIndex, 1);
+        timingRow.end.splice(aIndex, 1);
+        timingRow.oDepth.splice(bIndex, 1);
+        timingRow.oStack.splice(bIndex, 1);
+      }
+    }
+  }
+
+  // Compare neighboring stacks (a child, and parent). If both child and parent are
+  // platform code, then pop off the child, and shift the rest of the children stacks
+  // up.
+  for (let childDepth = stackTimingByDepth.length - 1; childDepth > 0; childDepth--) {
+    const parentDepth = childDepth - 1;
+    const parentTimingRow = stackTimingByDepth[parentDepth];
+    const childTimingRow = stackTimingByDepth[childDepth];
+
+    // Go through each stack frame at this depth.
+    for (let childTimingIndex = 0; childTimingIndex < childTimingRow.start.length; childTimingIndex++) {
+      // Find the parent frame.
+      const childStart = childTimingRow.start[childTimingIndex];
+      const childEnd = childTimingRow.end[childTimingIndex];
+      const parentTimingIndex = findParentTimingIndex(parentTimingRow, childStart, childEnd);
+      const childTimingRowLengthBefore = childTimingRow.start.length;
+
+      // Are both stacks from the platform?
+      if (childTimingRow.stack[childTimingIndex] === -1 &&
+          parentTimingRow.stack[parentTimingIndex] === -1) {
+
+        childTimingRow.start.splice(childTimingIndex, 1);
+        childTimingRow.end.splice(childTimingIndex, 1);
+        childTimingRow.stack.splice(childTimingIndex, 1);
+
+        collapseStackDownOneLevel(childDepth, childTimingIndex, childStart, childEnd);
```

Comment:
Wouldn't it be easier to do the platform stack collapsing in a pass that runs before the stackTimingByDepth pass?

In fact, I think we should replace the "JS only" filter in the call tree with something similar. So I recommend just using the JS-only filter funcStackInfo and not doing your own filtering. And then we can replace filterToJSOnly with filterOutPlatformData (or something like that) and change both places at the same time.